subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVcsMapping.java

```diff
@@ -19,18 +19,22 @@
 import com.google.common.base.Preconditions;
 import org.gradle.api.Action;
 import org.gradle.api.artifacts.component.ComponentSelector;
+import org.gradle.api.internal.initialization.ClassLoaderScope;
 import org.gradle.vcs.VersionControlSpec;
 
 public class DefaultVcsMapping implements VcsMappingInternal {
     private final ComponentSelector requested;
     private final VersionControlSpecFactory specFactory;
+    private ClassLoaderScope classLoaderScope;
```

Comment:
This feels like something we should contextualize the factory with so that this class doesn't know anything about classloader scopes and such.

Reply:
I checked if this would be easy to do.  The annoying thing here is the classLoaderScope we want is created here: https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java#L109

and I don't know how to fish that out so that the `VcsMappingFactory` or `VersionControlSpecFactory` is still visible at build scope or higher (`VcsMappingFactory` is needed by the `VcsDependencyResolver` to run the vcs mapping rules).

Reply:
That same classLoaderScope can be accessed via `SettingsInternal.getRootClassLoaderScope()`. HTH